src/js/media-controller.js

```diff
@@ -0,0 +1,309 @@
+/*
+  The <media-chrome> can contain the control elements
+  and the media element. Features:
+  * Auto-set the `media` attribute on child media chrome elements
+    * Uses the element with slot="media"
+  * Take custom controls to fullscreen
+  * Position controls at the bottom
+  * Auto-hide controls on inactivity while playing
+*/
+import { defineCustomElement } from './utils/defineCustomElement.js';
+import { propagateMedia, setAndPropagateMedia } from './media-chrome-html-element.js';
+import { Window as window, Document as document } from './utils/server-safe-globals.js';
+import { MEDIA_PLAY_REQUEST, MEDIA_PAUSE_REQUEST } from './media-ui-events.js';
+
+const template = document.createElement('template');
+
+template.innerHTML = `
+  <style>
+    :host {
+      box-sizing: border-box;
+      position: relative;
+
+      /* Position controls at the bottom  */
+      display: flex;
+      flex-direction: column-reverse;
+
+      /* Default dimensions */
+      width: 100%;
+      max-width: 720px;
+      height: 480px;
+      background-color: #000;
+    }
+
+    /* Safari needs this to actually make the element fill the window */
+    :host(:-webkit-full-screen) {
+      /* Needs to use !important otherwise easy to break */
+      width: 100% !important;
+      height: 100% !important;
+    }
+
+    /* Position the media element to fill the container */
+    ::slotted([slot=media]) {
+      position: absolute;
+      top: 0;
+      left: 0;
+      width: 100%;
+      height: 100%;
+    }
+
+    /* Hide controls when inactive and not paused */
+    #container ::slotted(*) {
+      opacity: 1;
+      transition: opacity 0.25s;
+      visibility: visible;
+    }
+
+    #container.inactive:not(.paused) ::slotted(*) {
+      opacity: 0;
+      transition: opacity 1s;
+    }
+  </style>
+  <slot name="media"></slot>
+  <div id="container">
+    <slot></slot>
+  </div>
+`;
+
+/*
+  Loop through child nodes and set the media[State] on every child.
+*/
```

Comment:
This is what handles passing state down, in general. Would probably be more convenient if it pierces the shadow dom also.